mlflow/projects/docker.py

```diff
@@ -77,21 +77,30 @@ def build_docker_image(work_dir, repository_uri, base_image, run_id):
         workdir=MLFLOW_DOCKER_WORKDIR_PATH,
     )
     build_ctx_path = _create_docker_build_ctx(work_dir, dockerfile)
-    with open(build_ctx_path, "rb") as docker_build_ctx:
-        _logger.info("=== Building docker image %s ===", image_uri)
-        client = docker.from_env()
-        image, _ = client.images.build(
-            tag=image_uri,
-            forcerm=True,
-            dockerfile=posixpath.join(_PROJECT_TAR_ARCHIVE_NAME, _GENERATED_DOCKERFILE_NAME),
-            fileobj=docker_build_ctx,
-            custom_context=True,
-            encoding="gzip",
-        )
-    try:
-        os.remove(build_ctx_path)
-    except Exception:
-        _logger.info("Temporary docker context file %s was not deleted.", build_ctx_path)
+    client = docker.from_env()
+    if rebuild:
+        with open(build_ctx_path, "rb") as docker_build_ctx:
+            _logger.info("=== Building docker image %s ===", image_uri)
+            image, _ = client.images.build(
+                tag=image_uri,
+                forcerm=True,
+                dockerfile=posixpath.join(_PROJECT_TAR_ARCHIVE_NAME, _GENERATED_DOCKERFILE_NAME),
+                fileobj=docker_build_ctx,
+                custom_context=True,
+                encoding="gzip",
+            )
+        try:
+            os.remove(build_ctx_path)
+        except Exception:
+            _logger.info("Temporary docker context file %s was not deleted.", build_ctx_path)
+    else:
+        if not client.images.list(name=base_image):
```

Comment:
Do we want to build the image if it can't be pulled or retrieved locally?

Reply:
To build an image, we need Dockerfile, but the project directory may not contain it. We can support this if there is a feature request.